cloud-sql-proxy/cloud-sql-proxy.sh

```diff
@@ -214,6 +226,50 @@ EOF
 
 }
 
+
+# Helper to run any command with default bdutil settings with Fibonacci
+# backoff. If all retries fail, returns last attempt's exit code.
+# Args: "$@" is the command to run.
+function run_with_retries() {
+  local retry_backoff=(1 1 2 3 5 8 13 21 34 55 89 144)
+  local -a cmd=("$@")
+  echo "About to run '${cmd[*]}' with retries..."
+
+  local update_succeeded=0
+  for ((i = 0; i < ${#retry_backoff[@]}; i++)); do
+    if "${cmd[@]}"; then
+      update_succeeded=1
+      break
+    else
+      local sleep_time=${retry_backoff[$i]}
+      echo "'${cmd[*]}' attempt $(( $i + 1 )) failed! Sleeping ${sleep_time}." >&2
+      sleep ${sleep_time}
+    fi
+  done
+
+  if ! (( ${update_succeeded} )); then
+    echo "Final attempt of '${cmd[*]}'..."
+    # Let any final error propagate all the way out to any error traps.
+    "${cmd[@]}"
+  fi
+}
+
+
+function configure_hive_warehouse_dir(){
+  run_with_retries run_validation
```

Comment:
Why does this require retries? (If it is necessary, please add a comment)

If this is just to block until the metastore comes back up, you may want to split restarting the metastore into its own method.

Also, does the invocation on line 205 need retries too?

Reply:
because if you point to an existing db as hive metastore it'll just succeed, otherwise it needs to wait for master 0 to create it first, otherwise when you restart hive-metastore service it tries to connect to the db instance that's not yet created and it'll fail. 

line 205 doesn't need retry, because it runs on master0 only, and it runs after the db instance is created(if needed)